APIClientTests用のStubを用意する

orzup opened the pull request:
@takuminnnn and @alotofwe

これまでは本番環境にリクエストを投げていたのですが、それではPOSTとかすると大変まずいので、Stubを用意しました。これにより、コメントアウトしていたAPIClientTestsを動かせるようになりました。

[OHHTTPStubs](https://github.com/AliSoftware/OHHTTPStubs) を使って、Fixturesに入っているJSONを返すようにつくってあります。

**確認してほしいこと**
- 自動ログインの状態でテストがグリーンなこと
- ログアウトした状態でテストがグリーンなこと

orzup committed:
OHHTTPStubsをインストール

orzup committed:
JSONベタ書きで用意していたレスポンスを渡すように修正

orzup committed:
FixturesにJSONデータを分離

orzup committed:
APIStubを各APIによって分岐させるように修正

orzup committed:
デバッグ用のコードを削除

orzup committed:
メソッドに抽出

orzup committed:
メソッド抽出

orzup committed:
不要なimportを削除

orzup committed:
ファイルの文末の空行が消えていたので修正

alotofwe commented on KakicoTests/Helpers/APIStub.swift:
```diff
@@ -0,0 +1,34 @@
+    private func stubJson(fileName: String) -> OHHTTPStubsResponse {
+        var tmp = fileName
+        if !fileName.hasSuffix(".json") {
+            tmp = fileName + ".json"
```
tmp... fileNameだということが分かりやすい変数名のほうがありがたいかもしれません... :eyes:

orzup replied:
いまいち代案が浮かばず `tmp` にしちゃったんだよなあ〜

alotofwe replied:
fileNameを直接編集してはいけないんだったっけ...?

orzup replied:
fileName は let なので無理なんだ :sob:

alotofwe replied:
https://developer.apple.com/library/prerelease/ios/documentation/Swift/Conceptual/Swift_Programming_Language/Functions.html

のページの真ん中らへんによると，引数を`var fileName: String`にすることでvarにできる予感...!

orzup replied:
すごい！よい！ 修正します！

alotofwe commented on KakicoTests/Fixtures/500.json:
```diff
@@ -0,0 +1,4 @@
+{
+    status: "500",
+    error: "Internal Server Error"
+}
```
500で返るjson，APIの種類によっても違うのだけれど，userの場合は
https://github.com/alotofwe/sample_app-1/blob/master/app/controllers/api/users_controller.rb#L155-L168
の通り，noticeにメッセージを入れています~
この場合は，notice: "Sorry, but something was wrong."かなあ

alotofwe replied:
正しいfixturesを作るために，APIのドキュメント強化の気運が高まっている...? :panda_face:

orzup replied:
https://157.7.190.148/api/users/999999 をたたくと下のようになるから、そのまま使ったのだけど、違ったのか…
![image](https://cloud.githubusercontent.com/assets/1712281/9895975/fd972ce4-5c70-11e5-8425-e2ee585fc3c0.png)

alotofwe replied:
https://github.com/alotofwe/sample_app-1/pull/55 のバグで
意図せぬ500が返っていました，すみません :bow: :bow: :bow:

orzup replied:
なるほど！じゃあ修正しておきます〜！

alotofwe commented:
APIの叩く先をlocalhost:3000にして，手元のsample_app-1のログを確認したのだけれど
まだ実際にAPIが叩かれている... :fearful:

orzup committed:
冗長な記述を修正

メソッドの引数もvarにできる

orzup committed:
Fixturesのメッセージを修正

alotofwe commented on KakicoTests/Helpers/APIStub.swift:
```diff
@@ -0,0 +1,33 @@
+            fileName = fileName + ".json"
+        }
+        return OHHTTPStubsResponse(fileAtPath:OHPathForFile(fileName, self.dynamicType)!,
+            statusCode:200, headers:["Content-Type":"application/json"])
```
500や404の場合でも，statusCodeが200で固定になっている...

orzup replied:
このstatusCodeは使っていない（jsonの中身の"status"で判断している）ので、何番でもいいんだよなあ。こういうときどうするのがいいのだろう？

alotofwe replied:
statusCodeも厳密にするとしたら，stubJsonの第二引数でstatusCodeを渡す...?

何も知らない人がテストを書こうとした時に，statusCodeがいつも200だと
正しく設定されているものだと思い込んで，意図しないテストを書いてしまうかもしれない...?

alotofwe commented on KakicoTests/Classes/Models/APIClientTests.swift:
```diff
@@ -1,36 +1,34 @@
+            onSuccess: { (user) -> Void in
+                XCTFail()
+            }, onFailure: { (error) -> Void in
+                XCTAssertNotNil(error)
```
https://github.com/pepabo/railstutorial-ios/blob/development/Kakico/Classes/Models/APIClient.swift#L8
のように，仕様として，errorにはstatus codeとheaderしか入らないのかあ...
せっかく`KakicoTests/Fixtures/500.json`を作っても，参照する手段が無い...!

ただ，落ちる時は適当なbodyしか返していないし「落ちた」ということだけ分かればいい気がするので
status codeを厳密にして，それで判断すればいいのかな...?

orzup replied:
メッセージをきちんと返すように修正しました〜！

orzup committed:
baseURLがlocalhostになったときもテストが走るように修正

orzup commented:
> APIの叩く先をlocalhost:3000にして,手元のsample_app-1のログを確認したのだけれど
> まだ実際にAPIが叩かれている... 

これ再現できなかった…特にログは流れなかったのだけれどもう一度確認してもらってもいいかな？ :disappointed:

orzup committed:
stubJsonにstatusCodeを渡すように変更

alotofwe commented:
> これ再現できなかった…特にログは流れなかったのだけれどもう一度確認してもらってもいいかな？

https://github.com/pepabo/railstutorial-ios/commit/d05e53f5548cc451ce336b5abdecb5c101175692 で直っていました! ありがとう~ :bow:

alotofwe commented:
あ，あとすみません，変なidの時に返ってくるstatus codeは404です...
APIのバグのせいで申し訳ない... :bow:

orzup committed:
不正なidへのアクセスした場合を分離

また、まだStubを用意してないAPIの場合、500を返すように変更しています。実際にはこの500にアクセスされることはありませんが、defaultを記述しないとエラーでビルドできないので書いています

orzup committed:
不正なuserIdでアクセスした場合、エラーのメッセージを返すように変更

alotofwe commented on KakicoTests/Helpers/APIStub.swift:
```diff
@@ -0,0 +1,38 @@
+            }, withStubResponse: { (request: NSURLRequest) -> OHHTTPStubsResponse in
+                switch request.URL!.path! {
+                case "/api/users/\(self.userId)": return self.stubJson("user", status: 200)
+                case "/api/users/\(self.illegalValue)": return self.stubJson("user_404", status: 404)
```
swiftのコード内に`_`があるのはなんとなく気になってしまうけれど
json fileだし，swiftの命名規則は適用させなくていいのかな... :cyclone:

orzup replied:
ファイルだからあんまり気にしてなかった…！

alotofwe commented:
LGTMです!
json fileの名前は気になっただけなので，お好きにして下さい~

ku00 commented:
ログイン/ログアウト状態でテストが無事通りました! :ok_woman: です!

orzup commented:
ふたりともレビューありがとうございました！ マージします〜！
